File: Chapter 04/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter04/section_4_2_downward_dependency/after/OrderRepository.java
package com.grokkingsoftwarearchitecture.chapter04.section_4_2_downward_dependency.after;

/**
 * THE ABSTRACTION (Interface).
 * ARCHITECTURE NOTE: In Java, we don't prefix with 'I'. The 
 * interface is the "pure" name of the role.
 */
public interface OrderRepository {
    void save(Order order);
}

File: Chapter 04/Python/section_4_2_downward_dependency/after/order_repository.py
from abc import ABC, abstractmethod
from .order import Order

class OrderRepository(ABC):
    """
    THE ABSTRACTION (INTERFACE).
    ARCHITECTURE NOTE: In Python, we use ABC to define a 
    formal contract. This is the 'Inversion' point. 
    The Service depends on this abstract role, not 
    a concrete database implementation.
    """
    @abstractmethod
    def save(self, order: Order):
        """
        Contract: Any subclass must implement this method.
        """
        pass